### PR TITLE
Initial TorchScript version of the core classes

### PR DESCRIPTION
**Pull request by Luthaf**
This is still missing a bunch of magic functions (`__getitem__`, …) and other convenience which exist on equistore-core, but should be at a good starting point for review!

----
:books: Documentation preview :books:: https://equistore--263.org.readthedocs.build/en/263/

### CODE REVIEW

File: equistore-core/include/equistore.hpp [hunk 1]
```diff
@@ -1294,7 +1318,7 @@ class TensorBlock final {
         return array;
     }
 
-    friend class TensorMap;
+    friend class equistore_torch::TensorBlockHolder;
```
Comment (agoscinski) [inline equistore-core/include/equistore.hpp:1321]: I don't understand the change here. Is this correct?

* doesn't this require to have equistore-torch also installed?
* why is the friend class TensorMap? Isn't TensorMap still needed as friend class?

File: equistore-core/tests/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -1,9 +1,10 @@
 cmake_minimum_required(VERSION 3.10)
 project(equistore-tests)
 
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-set(CMAKE_NO_SYSTEM_FROM_IMPORTED ON)
+if (${CMAKE_VERSION} VERSION_EQUAL 3.10)
```
Comment (agoscinski) [inline equistore-core/tests/cpp/CMakeLists.txt:4]: why is this needed? What if I have a higher cmake version is used? Don't you not need to set the CXX standard?

File: equistore-core/tests/c-api-install.rs [hunk 3]
```diff
@@ -45,22 +41,19 @@ fn check_c_api_build_install() {
...
+    assert!(status.success(), "failed to run test project cmake configuration");
+
+    // build the code, linking to equistore
+    let mut cmake_build = utils::cmake_build(&build_dir);
+    let status = cmake_build.status().expect("could not run cmake");
+    assert!(status.success(), "failed to run test project cmake build");
+
+    // run the executables
+    let mut ctest = utils::ctest(&build_dir);
+    let status = ctest.status().expect("could not run ctest");
```
Comment (agoscinski) [inline equistore-core/tests/c-api-install.rs:57]: note for me: arguments have been moved to utils/mod.rs

File: scripts/package-torch.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
```
Comment (agoscinski) [inline scripts/package-torch.sh:2]: Maybe some doc what this script does?

File: equistore-torch/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    endif()
+endif()
+
+find_package(equistore 0.1 CONFIG REQUIRED)
+
+# FindCUDNN.cmake distributed with PyTorch is a bit broken, so we have a
+# fixed version in `cmake/FindCUDNN.cmake`
+set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake;${CMAKE_MODULE_PATH}")
+
+find_package(Torch 1.11 REQUIRED)
```
Comment (agoscinski) [inline equistore-torch/CMakeLists.txt:36]: here you require 1.11 and in the toml you require 1.13

**Review by agoscinski — COMMENTED** [review]
What is the idea for people to install equistore-torch later on? Using a extra_require?

File: .github/workflows/python-tests.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
+name: Python tests
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply .github/workflows/python-tests.yml:1]: yes, I'm slowly rebasing all the branches as more of the required changes get in master

File: equistore-core/tests/cpp/CMakeLists.txt (hunk 2 above)
Comment (Luthaf) [reply equistore-core/tests/cpp/CMakeLists.txt:4]: We now set the C++ standard using target properties (which is only available since cmake 3.11). See `target_compile_features(catch PUBLIC cxx_std_11)` and similar for all targets

File: scripts/package-torch.sh (hunk 4 above)
Comment (Luthaf) [reply scripts/package-torch.sh:2]: 👍

File: equistore-torch/CMakeLists.txt (hunk 5 above)
Comment (Luthaf) [reply equistore-torch/CMakeLists.txt:36]: 👍

File: equistore-core/tests/cpp/CMakeLists.txt (hunk 2 above)
Comment (Luthaf) [reply equistore-core/tests/cpp/CMakeLists.txt:4]: Sorry, I though I was answering to a different PR ^^ This commit is actually part of #233, and will be replaced by the version in #267 once we merge it.

File: scripts/package-torch.sh (hunk 4 above)
Comment (Luthaf) [reply scripts/package-torch.sh:2]: Done in #233

File: equistore-torch/CMakeLists.txt (hunk 5 above)
Comment (Luthaf) [reply equistore-torch/CMakeLists.txt:36]: Done in #233

File: equistore-torch/src/labels.cpp [hunk 7]
```diff
@@ -51,32 +51,35 @@ static torch::Tensor initializer_list_to_tensor(
     return tensor.reshape({static_cast<int64_t>(count), static_cast<int64_t>(size)});
 }
 
-static std::vector<std::string> normalize_names(torch::IValue names) {
+std::vector<std::string> equistore_torch::details::normalize_names(torch::IValue names, std::string context) {
```
Comment (agoscinski) [inline equistore-torch/src/labels.cpp:54]: Will there be doc for this? To what do the names need to be normalized? First thought the context has to do something with the context manager of torch, but it is just for error context. Maybe we can also name it this way (e.g. `error_msg_prefix`)?

File: equistore-torch/include/equistore/torch/labels.hpp [hunk 8]
```diff
@@ -76,24 +89,51 @@ class EQUISTORE_TORCH_EXPORT LabelsHolder: public torch::CustomClassHolder {
...
+        return !labels_.has_value();
     }
+
+    /// Transform a view of Labels into owned Labels, which can be further given
+    /// to equistore functions. This does nothing if the Labels are already
+    /// owned.
+    LabelsHolder to_owned() const;
+
 private:
+    struct CreateView {};
```
Comment (agoscinski) [inline equistore-torch/include/equistore/torch/labels.hpp:123]: what is this struct `CreateView` doing? Could not find any usage of this

File: equistore-torch/src/labels.cpp [hunk 9]
```diff
@@ -158,60 +159,272 @@ TorchLabels LabelsHolder::create(
...
+    if (!labels_.has_value()) {
+        C10_THROW_ERROR(ValueError,
+            "can not call this function on Labels view, call to_owned first"
+        );
+    }
+
+    return labels_.value();
+}
+
+LabelsHolder LabelsHolder::to_owned() const {
```
Comment (agoscinski) [inline equistore-torch/src/labels.cpp:208]: Not exactly sure what this does. So `has_value` checks if is a "None" type (part from Optional type)? If it is not "None", then return itself. Otherwise it creates an LabelHolder with the names and values. I am not sure when this else case is used.

File: equistore-torch/include/equistore/torch/labels.hpp (hunk 8 above)
Comment (Luthaf) [reply equistore-torch/include/equistore/torch/labels.hpp:123]: leftover from development! I'll remove it

File: equistore-torch/src/labels.cpp (hunk 9 above)
Comment (Luthaf) [reply equistore-torch/src/labels.cpp:208]: If `LabelsHolder::labels_` is an `equistore::Labels`, we return that. If `LabelsHolder::labels_` is a nullopt (i.e. None), then we create a new equistore::Labels, and a new LabelsHolder wrapping it. 

`LabelsHolder::labels_` will be a nullopt when we create a column view for some Labels, with something like `labels[["center", "structure"]]`

I'll add some more explanations around this

File: equistore-torch/include/equistore/torch/labels.hpp (hunk 8 above)
Comment (Luthaf) [reply equistore-torch/include/equistore/torch/labels.hpp:129]: Actually, this is used in the constructor just below, to differentiate it from the main one. I've added a comment about that.

**Comment by Luthaf** [comment]
This should be ready for review and merge! For anyone wanting to give this a look, this PR is better reviewed one commit at the time.

**Comment by agoscinski** [comment]
Hello Guillaume,

Thanks for the work! I tried it out with the spherical expansion code from @frostedoyster and it worked after some changes.

The only problem I had with the core package is that the `__iter__` function is not available which is convenient to use. Do I understand correct that to include the `__iter__` function, we would need to do another change like this one https://github.com/pytorch/pytorch/pull/100724/ ?
In this case it is probably is not even possible because the output type is generic? I am also confused about where this `items()` function is implemented. I could only find the declaration.

Another point the `__repr__` behavior is different
```python
In [6]: tm_ref # equistore
Out[6]: 
TensorMap with 6 blocks
keys: ['a_i' 'lam' 'sigma']
         1     0      1
         6     0      1
         8     0      1
         1     1      1
         6     1      1
         8     1      1
```

```python
In [125]: tm_ref #equistore.torch
Out[125]: <torch.ScriptObject at 0x7f47cd423e70>
```
---
I am not sure if this is something that is part of this PR, but I wanted to mention it in case it is: When I tried to add equistore-torch and equistore-operations as setup requirements,
https://github.com/agoscinski/torch_spex/blob/cdac1549948250cd3ff407465bdc7478882d16e8/setup.py#L19-L32
I get a nonsensical dependency conflict on the same version.
```
ERROR: Cannot install equistore-operations==0.1.0.dev286 and equistore-torch==0.1.0.dev286 because these package versions have conflicting dependencies.
ERROR: Resolution Impossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
The conflict is caused by:
    equistore-operations 0.1.0.dev286 depends on equistore-core 0.1.0.dev286 (from /tmp/pip-install-6b03c972/equistore-operations_e14fe4088fca448786f5b9b3654f02b8/python/equistore-core)
    equistore-torch 0.1.0.dev286 depends on equistore-core 0.1.0.dev286 (from /tmp/pip-install-6b03c972/equistore-torch_a4828d7d7a304d0eb6b2b79c1aceaca7/python/equistore-core)
```
https://github.com/agoscinski/torch_spex/actions/runs/5171050634/jobs/9314351880

In the end I made it work by installing them separately in the tox file, and it worked fine.

**Comment by Luthaf** [comment]
> The only problem I had with the core package is that the __iter__ function is not available which is convenient to use. Do I understand correct that to include the __iter__ function, we would need to do another change like this one https://github.com/pytorch/pytorch/pull/100724 ?

No, we need much more than a simple change in torch: the `__iter__` function is returning a generator (i.e. it uses `yield` statements), but these are not supported by TorchScript, so we would have to add a significant feature to TorchScript. The plan is to replace `for key, block in tensor` with a more explicit `for key, block in tensor.items()`.

> Another point the __repr__ behavior is different

https://github.com/pytorch/pytorch/pull/100724 fixes this, but will only be available in torch 2.1

> I am not sure if this is something that is part of this PR, but I wanted to mention it in case it is: When I tried to add equistore-torch and equistore-operations as setup requirements,
https://github.com/agoscinski/torch_spex/blob/cdac1549948250cd3ff407465bdc7478882d16e8/setup.py#L19-L32
I get a nonsensical dependency conflict on the same version.

This is a known issue, but it is getting very annoying. I'll check if there is a solution, but I think the only way out is to make a release on PyPI

**Comment by agoscinski** [comment]
Thanks for the answers.

I encountered a problem when scripting LabelsView which you get when accessing a Labels with square brackets `__getitem__`
```python
import torch
from equistore.torch import Labels

@torch.jit.script
def foo():
    samples=Labels(
        names=["s_1", "s_2"], values=torch.IntTensor([[0, 0], [1, 1], [2, 2]])
    )
    return samples["s_1"].values
```
Error
```
Traceback (most recent call last):
  File "/home/alexgo/code/torch_spex/labelsview-scripting.py", line 5, in <module>
    def foo():
  File "/opt/anaconda3/envs/equitorch/lib/python3.10/site-packages/torch/jit/_script.py", line 1341, in script
    fn = torch._C._jit_script_compile(
RuntimeError: 
'Any' object has no attribute or method 'values'.:
  File "/home/alexgo/code/torch_spex/labelsview-scripting.py", line 9
        names=["s_1", "s_2"], values=torch.IntTensor([[0, 0], [1, 1], [2, 2]])
    )
    return samples["s_1"].values
           ~~~~~~~~~~~~~~~~~~~~ <--- HERE

```
This works
```python
import torch
from equistore.torch import Labels

@torch.jit.script
def foo():
    samples=Labels(
        names=["s_1", "s_2"], values=torch.IntTensor([[0, 0], [1, 1], [2, 2]])
    )
    return samples.values[:, samples.names.index("s_1")]
```

**Comment by Luthaf** [comment]
Yes, this is a known limitation: https://github.com/lab-cosmo/equistore/blob/f995ba2c479941cc4feff19f666ea73c83679a89/python/equistore-torch/equistore/torch/documentation.py#L251-L254

You can either use the `view()` function explicitly

```py
samples.view("s_1").values
```

or help the compiler refine types:

```py
view = samples["s_1"]
assert isinstance(view, Labels)
```

File: equistore-torch/src/array.cpp [hunk 10]
```diff
@@ -1,44 +1,95 @@
...
-#include <torch/torch.h>
+#include <torch/script.h>
 
 #include <equistore.hpp>
 
 #include "equistore/torch/array.hpp"
 
 using namespace equistore_torch;
 
+eqs_data_origin_t equistore_torch::TORCH_DATA_ORIGIN = 0;
```
Comment (agoscinski) [inline equistore-torch/src/array.cpp:12]: Can this get a bit more explanation? This is not clear to me how this constants decided?

File: equistore-torch/src/array.cpp [hunk 11]
```diff
@@ -1,44 +1,95 @@
...
+
+// eqs_data_origin registration in a thread-safe way through C++11 static
+// initialization of a class with a constructor
+struct EqsOriginRegistration {
```
Comment (agoscinski) [inline equistore-torch/src/array.cpp:16]: Can this get a bit more explanation? I don't understand why this thread-safe.

File: equistore-torch/src/tensor.cpp [hunk 12]
```diff
@@ -1,19 +1,432 @@
...
+        return {};
+    }
+
+    auto block = this->block_by_id(0);
+    auto n_dimensions = block->values().sizes().size();
+
+    return labels_names(block->as_equistore(), n_dimensions - 1);
+}
+
+std::vector<std::tuple<TorchLabelsEntry, TorchTensorBlock>> TensorMapHolder::items() {
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply equistore-torch/src/tensor.cpp:412]: Since this is a simple wrapper, I would rather re-implement it on the Python side directly. This reduces the number of functions/types in the shared library, reducing the cost to add new wrappers

File: equistore-torch/src/array.cpp (hunk 10 above)
Comment (Luthaf) [reply equistore-torch/src/array.cpp:12]: this is just initialized to 0 here (meaning "no origin"), and later allocated by the code in `EqsOriginRegistration`

File: docs/src/reference/torch/index.rst [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    from equistore import TensorMap, TensorBlock, Labels
+
+to
+
+.. code-block:: python
+
+    import equistore.torch as equistore
+    from equistore.torch import TensorMap, TensorBlock, Labels
+
+is 80% of the work required to make a model developed in Python --- without
```
Comment (agoscinski) [inline docs/src/reference/torch/index.rst:32]: If you only use equistore operations from core and equistore-operations, that will be True, but if you have custom equistore operations then not.

```suggestion
makes all the handling of equistore objects through equistore-operations compatible with TorchScript.
```

File: docs/src/reference/torch/index.rst [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+such as `Jax`_ arrays, `cupy`_ arrays, *etc.*.
+
+While this works great for a lot of use cases based on torch (defining models,
+training them with autograd, …), the TorchScript compiler is more restrictive in
+what code it accepts and does not support calling into arbitrary native code
+like the Python API does. The TorchScript version of equistore is here to fill
+this gap, teaching the TorchScript compiler what's going on inside equistore and
+how to translate code using equistore to TorchScript.
+
+Another different is that while the Python API supports storing data in multiple
```
Comment (agoscinski) [inline docs/src/reference/torch/index.rst:79]:
```suggestion
Another difference is that while the Python API supports storing data in multiple
```

File: equistore-torch/src/array.cpp (hunk 10 above)
Comment (Luthaf) [reply equistore-torch/src/array.cpp:12]: I'll add some more comment around this

File: docs/src/reference/torch/index.rst [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+ways, including storing it on GPU through torch Tensor, the metadata is always
+stored on CPU, inside the equistore shared library. The TorchScript API changes
+this to enable storing both the data and metadata on GPU, minimizing data
+transfer and making the models faster.
+
+If you only care about PyTorch, we would recommend using the TorchScript API
```
Comment (agoscinski) [inline docs/src/reference/torch/index.rst:85]:
```suggestion
If your code only needs to work with PyTorch, we would recommend using the TorchScript API
```

File: docs/src/reference/torch/index.rst [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+some level of overlap (it is possible to store data in torch ``Tensor`` with the
+Python API), so a big question is when/why should you use one or the other, and
+what's the difference between them.
+
+First, the Python and TorchScript API are separate to allow using the Python API
+without torch, for example in a pure numpy workflow, or with other array types
+such as `Jax`_ arrays, `cupy`_ arrays, *etc.*.
+
+While this works great for a lot of use cases based on torch (defining models,
+training them with autograd, …), the TorchScript compiler is more restrictive in
```
Comment (agoscinski) [inline docs/src/reference/torch/index.rst:73]:
```suggestion
training them with autograd, …), the `TorchScript`_ compiler is more restrictive in
```
added reference suggestion at the end

File: docs/src/reference/torch/index.rst [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+this to enable storing both the data and metadata on GPU, minimizing data
+transfer and making the models faster.
+
+If you only care about PyTorch, we would recommend using the TorchScript API
+from the start, to make sure you will be able to export your code to
+TorchScript. If you are not using PyTorch, or if you want to write code in an
+engine agnostic way, we recommend using the Python API.
+
+.. _Jax: https://jax.readthedocs.io/en/latest/
+.. _cupy: https://cupy.dev
```
Comment (agoscinski) [inline docs/src/reference/torch/index.rst:91]:
```suggestion
.. _cupy: https://cupy.dev
.. _TorchScript: https://pytorch.org/docs/stable/jit_language_reference.html#language-reference
```
